angelsrefining/prototypes/angels-functions.lua

```diff
@@ -1375,17 +1375,69 @@ function angelsmods.functions.remove_crafting_category(crafting_machine_type, cr
   end
 end
 
+local function box_equal(b1, b2)
+  if not (b1 and b2) then return false end
+
+  local function pos_equal(p1, p2)
+    if not (p1 and p2) then return false end
+
+    local p1x = p1.x or p1[1] or nil
+    local p2x = p2.x or p2[1] or nil
+    if not (p1x and p1x == p2x) then return false end
+
+    local p1y = p1.y or p1[2] or nil
+    local p2y = p2.y or p2[2] or nil
+
+    return (p1y and p1y == p2y)
+  end
+
+  if not pos_equal(b1.left_top or b1[1], b2.left_top or b2[1]) then return false end
+  return pos_equal(b1.right_bottom or b1[2], b2.right_bottom or b2[2])
+end
+-------------------------------------------------------------------------------
+-- MODIFY FAST_REPLACE_CATEGORY -----------------------------------------------
+-------------------------------------------------------------------------------
+function angelsmods.functions.set_fast_replace_category(crafting_machine_type, crafting_machine_name, next_upgrade)
+  --search for new category (if needed), skip if identical
+  if not data.raw[crafting_machine_type] then return end
+  local crafting_machine1 = data.raw[crafting_machine_type][crafting_machine_name]
+  if not crafting_machine1 then return end
+  local crafting_machine2 = data.raw[crafting_machine_type][next_upgrade]
+  if not crafting_machine2 then return end

```

Comment:
With the changes made in https://github.com/Arch666Angel/mods/pull/405/commits/547c05a69f1395276549a2efd87382d8b309e736, the argument `crafting_machine_name` or `next_upgrade` could be nil in the cases that the `crafting_machineX.next_upgrade` does not exist. In that case it will error trying to index with nil.

```suggestion
  local crafting_machine1 = crafting_machine_name and data.raw[crafting_machine_type][crafting_machine_name]
  if not crafting_machine1 then return end
  local crafting_machine2 = next_upgrade and data.raw[crafting_machine_type][next_upgrade]
  if not crafting_machine2 then return end
```